### PR TITLE
[RSDK-9475] - Update to use Image endpoint

### PR DESCRIPTION
**Pull request by seanavery**
### Description

This PR simply updates RDK dependency and uses new camera interface with `Image` defined instead of relying on video source wrapper. 

> NOTE: this does NOT remove pool as we still need it for efficient C to golang bytes transitions.

### Tests
Manually tested that Image endpoint successfully returns JPEG image and pool maintains integrity when polling over an extended period of time.
- viamrtsp h264 feed ✅ 
  - Image ✅ 
  - passthrough ✅ 
- viamrtsp h265 feed ✅ 
- viamrtsp mjpeg feed ✅ 
- reconfigures ✅ 
___

### TODO
- [x] Implement camera interface
- [x] Remove readerfunc and gostream wrapper handling
- [x] Image endpoint that returns JPEG image
- [x] Make pool work with Image
- [x] Get tests to pass with new interface
- [x] Handle MJPEG case
- [x] Create mime.go file for future mimetype handlers
- [x] Add props response
- [x] Make sure jpeg encode is as efficient as main
- [x] Make sure that passthrough still working

### CODE REVIEW

File: test/client.go [hunk 1]
```diff
@@ -52,14 +57,14 @@ func run() error {
...
-	if err != nil {
-		return err
-	}
-	_, _, err = stream.Next(ctx)
-	if err != nil {
-		return err
+
+	for range maxRetries {
+		_, err = camera.DecodeImageFromCamera(ctx, "image/jpeg", nil, ipCam)
+		if err == nil {
```
Comment (randhid) [inline test/client.go:63]: you know... sicne these are strings we could just implement YUV returns from viamrtsp right now

Comment (randhid) [reply test/client.go:63]: meant to comment on line 62

**Review by randhid — APPROVED** [review]

File: mime.go [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+	"image"
+	"image/jpeg"
+
+	"go.viam.com/rdk/components/camera"
+	rutils "go.viam.com/rdk/utils"
+)
+
+func encodeToJPEG(img image.Image) ([]byte, camera.ImageMetadata, error) {
+	buf := new(bytes.Buffer)
+	if err := jpeg.Encode(buf, img, nil); err != nil {
```
Comment (hexbabe) [inline mime.go:14]: I think in RDK we do quality 75

https://github.com/viamrobotics/rdk/blob/567b1abdc2f72008ec4f602c6e3af0ac3d6b3143/rimage/image_file.go#L323

File: rtsp_test.go [hunk 3]
```diff
@@ -74,11 +74,10 @@ func TestRTSPCamera(t *testing.T) {
 			defer imageTimeoutCancel()
 			var im image.Image
 			for imageTimeoutCtx.Err() == nil {
-				img, f, err := camera.ReadImage(imageTimeoutCtx, rtspCam)
+				img, err := camera.DecodeImageFromCamera(imageTimeoutCtx, "image/jpeg", nil, rtspCam)
```
Comment (hexbabe) [inline rtsp_test.go:77]: Could we import the mimetype enum from RDK here?

**Comment by hexbabe** [comment]
Also just in case can you include test results for rtp_passthrough

**Comment by hexbabe** [comment]
Could you also run

```
Makefile

profile-cpu:
	go test -v -cpuprofile cpu.prof -run "^TestRTSPCameraPerformance$$" -bench github.com/viam-modules/viamrtsp
	go tool pprof -top cpu.prof > cpu-profile.txt
	rm cpu.prof

profile-memory:
	go test -v -memprofile mem.prof -run "^TestRTSPCameraPerformance$$" -bench github.com/viam-modules/viamrtsp
	go tool pprof -top mem.prof > mem-profile.txt
	rm mem.prof
```

on main vs. on this feature branch and attach the reports

File: mime.go (hunk 2 above)
Comment (seanavery) [reply mime.go:14]: thanks good to know. looks like 75 is the default anyhow

https://pkg.go.dev/github.com/go-oss/image/jpeg#section-readme

File: rtsp_test.go (hunk 3 above)
Comment (seanavery) [reply rtsp_test.go:77]: yeah should use enum lookup in tests here -- will fix.

**Review by hexbabe — APPROVED** [review]

**Comment by seanavery** [comment]
@hexbabe nice profile test scripts!

Feature Branch CPU Profile:
[cpu-profile.txt](https://github.com/user-attachments/files/18156416/cpu-profile.txt)

Feature Branch Memory Profile:
[mem-profile.txt](https://github.com/user-attachments/files/18156420/mem-profile.txt)

Main CPU  Profile:
[cpu-profile.txt](https://github.com/user-attachments/files/18156442/cpu-profile.txt)

Main Memory Profile:
[mem-profile.txt](https://github.com/user-attachments/files/18156444/mem-profile.txt)


- Does not appear to be an appreciable discrepancy between feature and main
  - CPU test time not entirely reliable as I am running on a macbook with apps and such running

**Comment by hexbabe** [comment]
> @hexbabe nice profile test scripts!
> 
> Feature Branch CPU Profile: [cpu-profile.txt](https://github.com/user-attachments/files/18156416/cpu-profile.txt)
> 
> Feature Branch Memory Profile: [mem-profile.txt](https://github.com/user-attachments/files/18156420/mem-profile.txt)
> 
> Main CPU Profile: [cpu-profile.txt](https://github.com/user-attachments/files/18156442/cpu-profile.txt)
> 
> Main Memory Profile: [mem-profile.txt](https://github.com/user-attachments/files/18156444/mem-profile.txt)
> 
> * Does not appear to be an appreciable discrepancy between feature and main
>   
>   * CPU test time not entirely reliable as I am running on a macbook with apps and such running

This is interesting. Were there printfs in the feature test that weren't there on main? Seeing `     0.09s  7.63% 26.27%      0.19s 16.10%  fmt.(*pp).doPrintf`

**Comment by seanavery** [comment]
@hexbabe 
> This is interesting. Were there printfs in the feature test that weren't there on main? Seeing      0.09s  7.63% 26.27%      0.19s 16.10%  fmt.(*pp).doPrintf

Hm.. I do not think I am adding any extra logs. My best guess is something downstream from bumping `rdk/logging`. The only prints I see from tests are debug pool logs `Frame was put in pool. Len now: 1`.

More notes:
- somewhat hard to tell here if the jpeg encode is more or less efficient -- will try to get a better handle on this before merging.
```
      50ms  5.05% 34.34%       80ms  8.08%  image/jpeg.(*decoder).reconstructBlock
      50ms  5.05% 39.39%       50ms  5.05%  runtime.(*mspan).writeHeapBitsSmall
      40ms  4.04% 43.43%       40ms  4.04%  image/jpeg.fdct
      30ms  3.03% 46.46%       30ms  3.03%  image/jpeg.div (inline)
      30ms  3.03% 49.49%       30ms  3.03%  image/jpeg.idct
```
-  `70ms  7.07% 23.23%      150ms 15.15%  fmt.(*pp).handleMethods` is also concerning as it does not pop up in main branch.

**Comment by seanavery** [comment]
Also tried out going through rimage encode path which mostly does the same thing by encoding to jpeg.

[cpu-profile-jpeg.txt](https://github.com/user-attachments/files/18170983/cpu-profile-jpeg.txt)

[cpu-profile-riamge.txt](https://github.com/user-attachments/files/18170984/cpu-profile-riamge.txt)

**Comment by hexbabe** [comment]
I might just be reading this wrong

Screenshot 2024-12-17 at 4 06 41 PM
Screenshot 2024-12-17 at 4 06 53 PM

Pic 1 is feature. It's taking 650ms of cpu time. Pic 2 is main. It's taking over 1s of cpu time. Isn't this a performace regression?

**Comment by seanavery** [comment]
> Pic 2 is feature. It's taking 650ms of cpu time. Pic 1 is main. It's taking over 1s of cpu time. Isn't this a performace regression?

Yes -- just retested with latest main since the RDK bump and the profile is at 6 seconds now on main!

[cpu-profile-main-latest.txt](https://github.com/user-attachments/files/18171467/cpu-profile-main-latest.txt)

**Comment by hexbabe** [comment]
> > Pic 2 is feature. It's taking 650ms of cpu time. Pic 1 is main. It's taking over 1s of cpu time. Isn't this a performace regression?
> 
> Yes -- just retested with latest main since the RDK bump and the profile is at 750ms
> 
> [cpu-profile-main-latest.txt](https://github.com/user-attachments/files/18171467/cpu-profile-main-latest.txt)

oh ok sweet!

**Pull request merged**